fem/nonlininteg.hpp

```diff
@@ -88,6 +97,17 @@ class NonlinearFormIntegrator
        called. */
    virtual void AddMultPA(const Vector &x, Vector &y) const;
 
+   /// Method for partially assembled gradient action.
+   /** All arguments are E-vector. This method can be called only after the
+       method AssembleGradPA() has been called. */
+   virtual void AddMultGradPA(const Vector &g,
+                              const Vector &x, Vector &y) const;
+
+   /// Method for computing the diagonal of the gradient with partial assmebly.
+   /** All arguments are E-vector. This method can be called only after the
+       method AssembleGradPA() has been called. */
+   virtual void AssembleGradDiagonalPA(const Vector &x, Vector &diag) const;
```

Comment:
So, I've implemented in the above in my own branch of MFEM for ExaConstit as seen  [here](https://github.com/rcarson3/mfem/blob/exaconstit-dev/fem/nonlininteg.hpp#L102-L107). I guess I'm a little curious about what is the ```const Vector &g``` in ```AddMultGradPa``` and ```const Vector &x``` in ```AssembleGradDiagonalPA``` are supposed to be just because I at least didn't require these for my own implementations.

Reply:
I described the arguments in the comments, please take a look. The extra arguments are related to the comment in the PR description -- the PA-related `Setup` and `SetupGrad` of the `NonlinearForm` don't have an argument `x`, thus one needs to pass it it to integrators (by `g` in this case). Do you think they should have `x` as an argument?

Reply:
Okay, I see what you're talking about now.

> Do you think that `Setup` and `SetupGrad` should have `x` as an argument?

So, based on what I've done in `ExaConstit` I do feel that they should have a vector argument to them, which could be named something else to differentiate it from the vectors going into the `MultPA` and `MultGradPA` functions. The current `Setup` is a bit constraining in this manner and for example forces me to do something like this in my `NonlinearForm`-based class's [Mult](https://github.com/LLNL/ExaConstit/blob/exaconstit-dev/src/mechanics_operator.cpp#L279-L286).

One other change I would make to  `Setup` and `SetupGrad` is also have them be virtual functions on the `NonlinearForm` class. I know within `ExaConstit` my `Setup` is actually both the `Setup` and `SetupGrad` due to how my problems are formulated both of these are best done at the same time. I also make use of the `Setup` phase for all of my different assembly operations, since they all share the info obtained from it.